File: apps/AI-Keyboard/iOS/CherryPadKeyboard/KeyboardState.swift
import SwiftUI

/// Actions the SwiftUI keyboard views trigger. Abstracting these behind a protocol
/// (instead of referencing `KeyboardViewController` directly) keeps the views free
/// of any UIInputViewController / SDK dependency, so they compile in the Simulator
/// preview app for visual iteration.
protocol KeyboardActions: AnyObject {
    func insert(_ text: String)
    func deleteBackward()
    func newLine()
    func nextKeyboard()
    func runAction(_ task: KeyboardTask)
    func insertResult()
    func dismissResult()
}

enum KeyPlane { case letters, numbers, symbols }

/// Observable state shared between the controller and the SwiftUI layout.
final class KeyboardState: ObservableObject {
    @Published var plane: KeyPlane = .letters
    @Published var shifted = true
    @Published var needsFullAccess = false
    @Published var processing = false        // an AI action is running
    @Published var statusText: String?       // "Downloading model… 40%", "Thinking…"
    @Published var resultText: String?       // ready-to-insert AI result
    @Published var activeTask: KeyboardTask?  // which action produced the result
    @Published var banner: String?           // transient hint / error

    weak var controller: KeyboardActions?
}

File: apps/AI-Keyboard/Android/app/src/main/java/ai/zetic/demo/cherrypad/keyboard/KeyboardState.kt
package ai.zetic.demo.cherrypad.keyboard

import ai.zetic.demo.cherrypad.model.KeyboardTask
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

enum class KeyPlane { LETTERS, NUMBERS, SYMBOLS }

/** Observable keyboard UI state (mirrors iOS `KeyboardState`). */
class KeyboardState {
    var plane by mutableStateOf(KeyPlane.LETTERS)
    var shifted by mutableStateOf(true)          // starts uppercase, like iOS
    var processing by mutableStateOf(false)
    var statusText by mutableStateOf<String?>(null)
    var resultText by mutableStateOf<String?>(null)
    var activeTask by mutableStateOf<KeyboardTask?>(null)
    var banner by mutableStateOf<String?>(null)
}

/** Text-editing + AI actions the keyboard view delegates to the IME service. */
interface KeyboardActions {
    fun insert(text: String)
    fun deleteBackward()
    fun newLine()
    fun nextKeyboard()
    fun runAction(task: KeyboardTask)
    fun insertResult()
    fun dismissResult()
}
